### PR TITLE
Additional dbwrapper tests

### PR DESCRIPTION
**Pull request by str4d**
Cherry-picked from the following upstream PRs:

- bitcoin/bitcoin#7992
- bitcoin/bitcoin#9867
  - Only the commit affecting dbwrapper tests
- bitcoin/bitcoin#9610
  - Only the change affecting dbwrapper tests
- bitcoin/bitcoin#10844

### CODE REVIEW

File: src/test/dbwrapper_tests.cpp [hunk 1]
```diff
@@ -116,5 +116,125 @@ BOOST_AUTO_TEST_CASE(dbwrapper_iterator)
...
+            BOOST_CHECK_EQUAL(key, x);
+            BOOST_CHECK_EQUAL(value, x*x);
+            it->Next();
+        }
+        BOOST_CHECK(!it->Valid());
+    }
+}
+
+struct StringContentsSerializer {
+    // Used to make two serialized objects the same while letting them have a different lengths
```
Comment (daira) [inline src/test/dbwrapper_tests.cpp]: s/a different lengths/different lengths/

File: src/test/dbwrapper_tests.cpp [hunk 2]
```diff
@@ -73,7 +74,7 @@ BOOST_AUTO_TEST_CASE(dbwrapper_batch)
         BOOST_CHECK(dbw.Read(key2, res));
         BOOST_CHECK_EQUAL(res.ToString(), in2.ToString());
 
-        // key3 never should've been written
+        // key3 should've never been written
```
Comment (daira) [inline src/test/dbwrapper_tests.cpp]: I could nitpick and say that this should be "key3 should never have been written", but I'll spare you that :-)

File: src/test/dbwrapper_tests.cpp [hunk 3]
```diff
@@ -194,7 +190,7 @@ BOOST_AUTO_TEST_CASE(iterator_string_ordering)
     CDBWrapper dbw(ph, (1 << 20), true, false);
     for (int x=0x00; x<10; ++x) {
         for (int y = 0; y < 10; y++) {
-            sprintf(buf, "%d", x);
+            snprintf(buf, sizeof(buf), "%d", x);
```
Comment (daira) [inline src/test/dbwrapper_tests.cpp]:
```
int n = snprintf(buf, sizeof(buf), "%d", x);
assert(n > 0 && n < sizeof(buf));
```

Comment (daira) [reply src/test/dbwrapper_tests.cpp]: And similarly for the two ``snprintf``s below.

**Comment by str4d** [comment]
Implemented @daira's comment.

@zkbot r+

**Comment by str4d** [comment]
Gah, failed due to a transient RPC test timeout.

@zkbot retry

**Pull request merged**